### DISCUSSION TITLE
How to playback recordings with a gap?

### DISCUSSION THREAD

**Discussion by moenodedev** [bluenviron/mediamtx#3243] [Questions]
### Question

Can the playback server stitch 2 fmp4 segments together that have a gap in between? I have a segment from 9-9:10 and another segment from 9:15-9:30.  I would like to play the range from 9:05-9:20 as one video skipping the 5 minute gap.

**Comment by aler9** [comment]
hello, this cannot be done automatically but you can query the playback server in order to find out available timespans for a certain source:

```
curl http://localhost:9996/list?path=mypath
```

obtaining:

```json
[
  {
    "start": "2006-01-02T15:04:05Z07:00",
    "duration": "60.0"
  },
  {
    "start": "2006-01-02T15:07:05Z07:00",
    "duration": "32.33"
  }
]
```

then pull the individual segments:

```
curl -o segment1.mp4 http://localhost:9996/get?path=mypath&start=2006-01-02T15%3A04%3A05Z07%3A00
```

then merge then together:

```
ffmpeg -i segment1.mp4 -c copy intermediate1.ts
ffmpeg -i segment2.mp4 -c copy intermediate2.ts
ffmpeg -i "concat:intermediate1.ts|intermediate2.ts" -c copy output.mp4
```

**Reply by moenodedev** [reply]
Thanks for the info 😄 We'll try this soon.

We created our own playback server without the playback API because listing fmp4 segments is slow and the video doesn't play in browsers on macOS.

We're running nginx. A script fetches and parses `segments.txt` from `/$path/segments.txt` then creates an HLS playlist dynamically.

If a user says they want to view footage between 8:10 and 8:15pm we find all segments between 8:09 and 8:15 since the segment may start before the start of the minute. The clip length doesn't need to be exact. 

These are our recording settings that save HLS segments. The `runOnRecordSegmentComplete` hook saves the filename and duration to segments.txt.
```
      "record": true,
      "recordPath": "./recordings/%path/%Y-%m-%d_%H-%M-%S-%f",
      "recordFormat": "mpegts",
      "recordPartDuration": "1s",
      "recordSegmentDuration": "10s",
      "runOnRecordSegmentComplete": "sh -c \"echo $MTX_SEGMENT_PATH $MTX_SEGMENT_DURATION >>/recordings/$MTX_PATH/segments.txt\""
```

segments.txt
```
./recordings/front/2024-05-22_00-23-37-678440.ts 10.003611
./recordings/front/2024-05-22_00-23-47-668940.ts 10.043611
./recordings/front/2024-05-22_00-23-57-666210.ts 10.043611
./recordings/front/2024-05-22_00-24-07-662915.ts 12.003611
./recordings/front/2024-05-22_00-24-19-667822.ts 10.053611
./recordings/front/2024-05-22_00-24-29-657168.ts 10.053611
./recordings/front/2024-05-22_00-24-39-661342.ts 12.013611
./recordings/front/2024-05-22_00-24-51-657152.ts 10.063611
./recordings/front/2024-05-22_00-25-01-654187.ts 10.063611
./recordings/front/2024-05-22_00-25-11-653595.ts 9.983611
./recordings/front/2024-05-22_00-25-21-644015.ts 12.024000
./recordings/front/2024-05-22_00-25-33-640779.ts 9.993611
./recordings/front/2024-05-22_00-25-43-638905.ts 9.993611
./recordings/front/2024-05-22_00-25-53-635906.ts 10.033611
```

[a comment by an automated account is not shown]
